Fix json payload error which causes recursion error in the front end …

upendrakumbham opened the pull request:
Issue reference: [Recursion error](https://github.com/ebi-gene-expression-group/atlas-web-single-cell/issues/278)

Getting recursion error in the front end because of the JSON response error for the metadata 'Gallus gallus'. In the JSON response tree, one element generates the same value for the "ID" and "Parent" attributes.

Recursion error URL:  https://www.ebi.ac.uk/gxa/sc/search/metadata/Gallus%20gallus

Endpoint - ` https://www.ebi.ac.uk/gxa/sc/json/cell-type-wheel/Gallus gallus`

Fix: Tweaked a code to generate JSON payload attribute values "ID" and "Parent" are always different.

upendrakumbham committed:
Fix json payload error which causes recursion error in the front end - json tree attribute values of 'id' and 'parent' should not be the same

upendrakumbham committed:
Fix json payload error which causes recursion error in the front end - json tree attribute values of 'id' and 'parent' should not be the same

alfonsomunozpomer reviewed (changes requested):
Please add a test that fails with the faulty code and passes after adding your fix.

upendrakumbham committed:
Add a unit test for the recursion bug

upendrakumbham committed:
Merge branch 'bugFix/recursion-error-when-searching-for-a-metadata-gallus-gallus' of github.com:ebi-gene-expression-group/atlas-web-single-cell into bugFix/recursion-error-when-searching-for-a-metadata-gallus-gallus
* 'bugFix/recursion-error-when-searching-for-a-metadata-gallus-gallus' of github.com:ebi-gene-expression-group/atlas-web-single-cell:
  Fix json payload error which causes recursion error in the front end - json tree attribute values of 'id' and 'parent' should not be the same

upendrakumbham committed:
Add a test to test pre and post bugfix functionality

upendrakumbham committed:
Merge branch 'develop' into bugFix/recursion-error-when-searching-for-a-metadata-gallus-gallus

alfonsomunozpomer commented on app/src/test/java/uk/ac/ebi/atlas/search/metadata/HighchartsSunburstAdapterTest.java:
```diff
@@ -0,0 +1,110 @@
+                                species1OrganismPart2ExperimentAccession)));
+    }
+
+    @DisplayName("This unit test use the randomised input test data and test the functionality of the wheel")
```
Rephrase without words such as “This unit test...” or “This test...” (see below). This is more of a code comment rather than a test description, since it mentions implementation details. The `@DisplayName` should be a human-readable firendlier version of the test method; for instance, something like:
> Cell type wheel should not have the same value for the ID and its parent

alfonsomunozpomer commented on app/src/test/java/uk/ac/ebi/atlas/search/metadata/HighchartsSunburstAdapterTest.java:
```diff
@@ -0,0 +1,110 @@
+    @DisplayName("This unit test use the randomised input test data and test the functionality of the wheel")
+    @Test
+    void cellTypeWheelSunburstResultObjectShouldNotContainIdAndParentAttributeValuesSame() {
+        ImmutableSet<ImmutableMap<String, ?>> result = subject.getCellTypeWheelSunburst(metadataSearchTerm,cellTypeWheelServiceMock
```
You can use `var` here.

alfonsomunozpomer commented on app/src/test/java/uk/ac/ebi/atlas/search/metadata/HighchartsSunburstAdapterTest.java:
```diff
@@ -0,0 +1,110 @@
+                                ImmutablePair.of(
+                                        ImmutableList.of("Gallus gallus"), ("E-GEOD-89910"))));
+
+        ImmutableSet<ImmutableMap<String, ?>> result = subject.getCellTypeWheelSunburst("Gallus gallus",cellTypeWheelServiceMock
```
Use `var` again.

alfonsomunozpomer commented on app/src/test/java/uk/ac/ebi/atlas/search/metadata/HighchartsSunburstAdapterTest.java:
```diff
@@ -0,0 +1,110 @@
+
+    @Test
+    @DisplayName("This test uses hardcoded mock test data produced by the public server and tests the pre and post-bugfix code precisely to make sure that bug will not reappear again." +
+            "Both unit tests look the same, the only difference is in the input test data.")
```
Again, these are code comments. A unit test should’t mention other tests for understanding its meaning.

alfonsomunozpomer reviewed (changes requested):
Some changes in wording and style. Other than that it’s mostly ok.

alfonsomunozpomer reviewed (changes requested):
Some changes in wording and style. Other than that it’s mostly ok.

upendrakumbham committed:
Merge branch 'develop' into bugFix/recursion-error-when-searching-for-a-metadata-gallus-gallus

upendrakumbham committed:
Fix review comments

upendrakumbham committed:
Merge branch 'develop' into bugFix/recursion-error-when-searching-for-a-metadata-gallus-gallus

alfonsomunozpomer commented on app/src/test/java/uk/ac/ebi/atlas/search/metadata/HighchartsSunburstAdapterTest.java:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    @Test
+    void ShouldThrowAssertionErrorIfWheelResultIdAndParentAttributeValuesAreSame() {
```
Whenever a test fails, it throws an assertion error/exception, so this shouldn’t be part of what you’re testing. In other words, a test that is named:
```java
@Test
void throwsAssertionErrorIfTheSubjectWorksUnderSomeCondition() {
// ...
}
```

Should be renamed to:
```java
void subjectWorksUnderSomeCondition() {
// ...
}
```

One last observation: what’s the difference between this test and the previous one? What are you testing in each case? That should be your test name.

alfonsomunozpomer reviewed (changes requested):
Some clarification for the tests required.

upendrakumbham committed:
Rename method names and fix assertions as per review comments

alfonsomunozpomer commented on app/src/test/java/uk/ac/ebi/atlas/search/metadata/HighchartsSunburstAdapterTest.java:
```diff
@@ -75,18 +76,17 @@ void setUp() {
 
     @Test
-    void ShouldThrowAssertionErrorIfWheelResultIdAndParentAttributeValuesAreSame() {
+    void failsIfIdAndParentFieldValuesAreSame() {
```
Again, “fails” isn’t what you’re testing. Your test name describes the conditions and/or expectations of good behaviour. If it helps, @ke4 sent this a while ago to our Slack channel: https://osherove.com/blog/2005/4/3/naming-standards-for-unit-tests.html.

alfonsomunozpomer committed:
Chore – Marker-genes-only search (#347)

* Chore – Remove experiment design (#316)

* Don’t render the Experiment Design tab

* Update atlas-web-core module

* Update atlas-web-core

* Update atlas-web-core

* Fix tests

* Update atlas-web-core

* Enable K8s SolrCloud scaler

* Update atlas-web-core module

* Update atlas-web-core module

* Update atlas-web-core module

* Update atlas-web-core module

* Update atlas-web-core module

* Remove ExperimentDesignTable

* Remove experiment design JSON object

* Build this branch

* Create production bundles unless branch is develop

* Chore – Gene search optimization (#333)

* DRY code

* Use DocValues field instead of the regular one and use .returnAllDocs() instead of a hard-coded “magic” number

As a sidenote, I find it really strange that Solr wasn’t complaining when we were using bioentity_identifier as the sort field (I suspect it was ignored)

* Change varargs to an immutable collection to avoid unnecessary wrapping-unwrapping of arrays to ImmutableSet

* Take times to find out where most of the gene search time is spent

* Change human Ensembl ID example to a lighter one and set Homo sapiens as default instead of all species

* Avoid APT 404 errors in Docker containers

* Add authentication to Tomcat container

* Update atlas-web-core module

* Use autoincrement syntax in for loop

* Change test to use only marker genes

* Use for instead of while

* Chore – Gene search optimization (#335)

* DRY code

* Use DocValues field instead of the regular one and use .returnAllDocs() instead of a hard-coded “magic” number

As a sidenote, I find it really strange that Solr wasn’t complaining when we were using bioentity_identifier as the sort field (I suspect it was ignored)

* Change varargs to an immutable collection to avoid unnecessary wrapping-unwrapping of arrays to ImmutableSet

* Take times to find out where most of the gene search time is spent

* Change human Ensembl ID example to a lighter one and set Homo sapiens as default instead of all species

* Avoid APT 404 errors in Docker containers

* Add authentication to Tomcat container

* Update atlas-web-core module

* Use autoincrement syntax in for loop

* Change test to use only marker genes

* Use for instead of while

* Simplify parsing of properties in Gradle tasks

* Add dev-test.env to avoid destroying pre-populated volumes in stopped containers running Postgres

* Add support of Solr authentication in integration tests

* Update atlas-web-core module

* Tweak test to use a human experiment

* Update atlas-web-core module

* Update atlas-web-core module

* Cherrypick Solr health endpoint fix (#338)

* Update Mockito to version 5.2

Mockito 2 throws if you try to mock a final method

* Set SCHEMA_VERSION to latest

* ./stop-and-remove-containers.sh brings down the test environment

* Fix matcher in test

With Mockito v2 `any(string.class)` passes, but it fails after updating Mockito to v5

* Clean code and remove lenient behaviour in Mockito

* Correctly express dependencies of the Gradle container

It should wait for Flyway, which in turn waits for Postgres

* cherry-pick atlas-web-core changes related to the health-endpoint

* go back to previous schema version (not Anndata yet)

* Update atlas-web-core release/202302 branch with cherry-picked commits from the health endpoint fix

* Fix GeneSearchServiceTest

---------

Co-authored-by: Alfonso Muñoz-Pomer Fuentes <amunoz@ebi.ac.uk>

* Change controller to work with new signature of GeneSearchService method

* Fix integration tests (part of https://github.com/ebi-gene-expression-group/atlas-web-single-cell/pull/346)

* Update atlas-web-core

* Fix more integration tests (part of https://github.com/ebi-gene-expression-group/atlas-web-single-cell/pull/346)

* Properly split `/json/gene-search` and `/json/search` tests

---------

Co-authored-by: Károly Erdős <2183410+ke4@users.noreply.github.com>

alfonsomunozpomer committed:
Chore – Marker genes only search (again) (#348)

* Chore – Remove experiment design (#316)

* Don’t render the Experiment Design tab

* Update atlas-web-core module

* Update atlas-web-core

* Update atlas-web-core

* Fix tests

* Update atlas-web-core

* Enable K8s SolrCloud scaler

* Update atlas-web-core module

* Update atlas-web-core module

* Update atlas-web-core module

* Update atlas-web-core module

* Update atlas-web-core module

* Remove ExperimentDesignTable

* Remove experiment design JSON object

* Build this branch

* Create production bundles unless branch is develop

* Chore – Gene search optimization (#333)

* DRY code

* Use DocValues field instead of the regular one and use .returnAllDocs() instead of a hard-coded “magic” number

As a sidenote, I find it really strange that Solr wasn’t complaining when we were using bioentity_identifier as the sort field (I suspect it was ignored)

* Change varargs to an immutable collection to avoid unnecessary wrapping-unwrapping of arrays to ImmutableSet

* Take times to find out where most of the gene search time is spent

* Change human Ensembl ID example to a lighter one and set Homo sapiens as default instead of all species

* Avoid APT 404 errors in Docker containers

* Add authentication to Tomcat container

* Update atlas-web-core module

* Use autoincrement syntax in for loop

* Change test to use only marker genes

* Use for instead of while

* Chore – Gene search optimization (#335)

* DRY code

* Use DocValues field instead of the regular one and use .returnAllDocs() instead of a hard-coded “magic” number

As a sidenote, I find it really strange that Solr wasn’t complaining when we were using bioentity_identifier as the sort field (I suspect it was ignored)

* Change varargs to an immutable collection to avoid unnecessary wrapping-unwrapping of arrays to ImmutableSet

* Take times to find out where most of the gene search time is spent

* Change human Ensembl ID example to a lighter one and set Homo sapiens as default instead of all species

* Avoid APT 404 errors in Docker containers

* Add authentication to Tomcat container

* Update atlas-web-core module

* Use autoincrement syntax in for loop

* Change test to use only marker genes

* Use for instead of while

* Simplify parsing of properties in Gradle tasks

* Add dev-test.env to avoid destroying pre-populated volumes in stopped containers running Postgres

* Add support of Solr authentication in integration tests

* Update atlas-web-core module

* Tweak test to use a human experiment

* Update atlas-web-core module

* Update atlas-web-core module

* Cherrypick Solr health endpoint fix (#338)

* Update Mockito to version 5.2

Mockito 2 throws if you try to mock a final method

* Set SCHEMA_VERSION to latest

* ./stop-and-remove-containers.sh brings down the test environment

* Fix matcher in test

With Mockito v2 `any(string.class)` passes, but it fails after updating Mockito to v5

* Clean code and remove lenient behaviour in Mockito

* Correctly express dependencies of the Gradle container

It should wait for Flyway, which in turn waits for Postgres

* cherry-pick atlas-web-core changes related to the health-endpoint

* go back to previous schema version (not Anndata yet)

* Update atlas-web-core release/202302 branch with cherry-picked commits from the health endpoint fix

* Fix GeneSearchServiceTest

---------

Co-authored-by: Alfonso Muñoz-Pomer Fuentes <amunoz@ebi.ac.uk>

* Change controller to work with new signature of GeneSearchService method

* Fix integration tests (part of https://github.com/ebi-gene-expression-group/atlas-web-single-cell/pull/346)

* Update atlas-web-core

* Fix more integration tests (part of https://github.com/ebi-gene-expression-group/atlas-web-single-cell/pull/346)

* Properly split `/json/gene-search` and `/json/search` tests

* Enable test that checks that experiments with marker genes for inferred cell types have the necessary files

* Clean DataFileHub integration tests

* Update atlas-web-core module

* Add ExperimentCellCountDaoImpl to match merged develop branch in atlas-web-core

* Restore @WebAppConfiguration to test that wants a ServletContext

* Update atlas-web-core module

---------

Co-authored-by: Károly Erdős <2183410+ke4@users.noreply.github.com>

alfonsomunozpomer committed:
Enable atlas-web-core integration tests in Single Cell Expression Atlas Gradle container (#346)

* Update Mockito to version 5.2

Mockito 2 throws if you try to mock a final method

* Set SCHEMA_VERSION to latest

* ./stop-and-remove-containers.sh brings down the test environment

* Update atlas-web-core module

* Fix matcher in test

With Mockito v2 `any(string.class)` passes, but it fails after updating Mockito to v5

* Clean code and remove lenient behaviour in Mockito

* Correctly express dependencies of the Gradle container

It should wait for Flyway, which in turn waits for Postgres

* Wrap all varibales in curly braces, for consistency

* Add utility function that outputs a message with a 10 seconds countdown you can skip or cancel

* Change Gradle caches task to create Docker Compose volumes

gradle-cache

* Change volumes task to create Docker Compose volumes

* Change Postgres task to create Docker Compose volumes

* Change Postgres task to create Docker Compose volumes

* Change Solr task to create Docker Compose volumes

* Fix Solr task to sign BioSolr JAR package with a secret

* Reorganise Postgres Docker Compose YAML file to follow the same structure as the others

* Parameterise Solr container names and volumes with PROJECT_NAME (scxa or gxa)

* Move webapp-properties to be together with docker-compose-tomcat.yml and parameterise Tomcat YAML with PROJECT_NAME

* Add YAML and script to build and deploy the web app

* Add environment variables to Flyway test service

* Remove accidentally pasted duplicate block in Postgres script

* Add -r parameter to usage in Solr script

* Create magetab directory in the exp (as per the volumes section)

* Remove hardcoded scxa prefix to Solr and ZooKeeper containers

* Add conditions to Gradle container dependencies

* Remove unnecessary dev-test.env file

* Tweak scripts to support different .env file and project name depending on the subproject for which tests are running

* Remove stale import

* Add implementation for ExperimentCellCountDao in app subproject

* Fix bean injection for tests

* Remove stale imports

* Do not hardcode paths, let Gradle dictate where files are located

* Reword comment

* Remove unused provider

* Update atlas-web-core

* Fix service dependency to reuse the Gradle container

* Update atlas-web-core submodule

* Improve documentation

* Update execute-single-test.sh to use dev.env file

* Replace magic words with random values

* Refactor EuropePMCClient to build URLs with a library; reimplement tests with mock server

* Update atlas-web-core module

* Update atlas-web-core module

* Update atlas-web-core module

* Update atlas-web-core to merged commit

* Remove environment variables that are passed via --env-file

* Remove environment variables that are passed via --env-file (Postgres edition)

* Remove environment variables that are passed via --env-file (Postgres stub edition)

* Declare POSTGRES_* variables to have them set in the configuration evironment

upendrakumbham committed:
Rename tests

upendrakumbham committed:
Merge branch 'develop' into bugFix/recursion-error-when-searching-for-a-metadata-gallus-gallus

alfonsomunozpomer reviewed (approved):
All good!
